### ISSUE TITLE
Function does not optimize, but reports no warnings in the debug log

### ISSUE THREAD

**Issue by carbonjesse** [flexport/reflective-bind#32]
The following function does not hoist and optimize, although it reports a successful optimization when running the transform in debug mode.

```
render() {
  return (<View onClick={ ({propA, propB}) => this.setState({data: {propA, propB}}) } </View>)
}
```

However the library will do the optimization when I make the following change.

```
doSetState(propA, propB) { this.setState({data:{ propA, propB}}); }

render() {
  return (<View onClick={ ({propA, propB}) => this.doSetState(propA, propB) } </View>)
}
```

Thanks

**Comment by dounan** [comment]
Thanks for bringing up this issue! I'll take a look at it over this weekend.

**Comment by dounan** [comment]
@carbonjesse I tried out the examples you gave, and could not reproduce the issue. Both examples transform correctly for me.

### this.setState

#### Original code
```jsx
import * as React from "react";

function View(props) {
  return null;
}

class MyComponent extends React.Component {
  render() {
    return (
      <View
        onClick={({propA, propB}) => this.setState({data: {propA, propB}})}
      />
    );
  }
}
```

#### Transformed code
```jsx
import { babelBind as _testBind } from \\"../../src\\";

function _testBBHoisted({
  propA,
  propB
}) {
  return this.setState({
    data: {
      propA,
      propB
    }
  });
}

import * as React from \\"react\\";

function View(props) {
  return null;
}

class MyComponent extends React.Component {
  render() {
    return <View onClick={_testBind(_testBBHoisted, this)} />;
  }
}
```

### doSetState

#### Original code
```jsx
import * as React from "react";

function View(props) {
  return null;
}

class MyComponent extends React.Component {
  doSetState(propA, propB) {
    this.setState({data: {propA, propB}});
  }

  render() {
    return <View onClick={({propA, propB}) => this.doSetState(propA, propB)} />;
  }
}
```

#### Transformed code
```jsx
import { babelBind as _testBind } from \\"../../src\\";

// @flow
function _testBBHoisted({
  propA,
  propB
}) {
  return this.doSetState(propA, propB);
}

import * as React from \\"react\\";

function View(props) {
  return null;
}

class MyComponent extends React.Component {
  doSetState(propA, propB) {
    this.setState({
      data: {
        propA,
        propB
      }
    });
  }

  render() {
    return <View onClick={_testBind(_testBBHoisted, this)} />;
  }
}
```

**Comment by carbonjesse** [comment]
OK! I did a bit more digging @dounan and I think I figured out what's going on.

My example was missing a bit of detail that turns out to be important. Below is an updated example that shows where things are not working.

### Does Not Work

**Original Code**
`import * as React from "react";

function View(props) {
  return null;
}

class MyComponent extends React.Component {
  const {data} = this.state;
  render() {
    return (
      <View
        onClick={({propA, propB}) => this.setState({data: {propA, propB}})}
      />
    );
  }
}`

**Transformed code**

`import { babelBind as _testBind } from \\"../../src\\";

function _testBBHoisted({
  propA,
  propB
}, data) {
  return this.setState({
    data: {
      propA,
      propB
    }
  });
}

import * as React from \\"react\\";

function View(props) {
  return null;
}

class MyComponent extends React.Component {
  const {data} = this.state;
  render() {
    return <View onClick={_testBind(_testBBHoisted, this, data)} />;
  }
}`

***Notice how the library wants to pass through the 'data' variable, even though data is only being used as a key.***

### This works

**Original Code**
`import * as React from "react";

function View(props) {
  return null;
}

class MyComponent extends React.Component {
  const {data} = this.state;
  render() {
    return (
      <View
        onClick={({propA, propB}) => this.setState({"data": {propA, propB}})}
      />
    );
  }
}`

**Transformed code**

`import { babelBind as _testBind } from \\"../../src\\";

function _testBBHoisted({
  propA,
  propB
}) {
  return this.setState({
    data: {
      propA,
      propB
    }
  });
}

import * as React from \\"react\\";

function View(props) {
  return null;
}

class MyComponent extends React.Component {
  const {data} = this.state;
  render() {
    return <View onClick={_testBind(_testBBHoisted, this)} />;
  }
}`

***In this second case, notice how it does not confuse the key 'data' with the variable, thus is able to hoist more efficiently and won't re-render the subcomponent.

Thanks

**Comment by dounan** [comment]
Which version of `reflective-bind` are you running? I'm currently debugging with the latest version of master.

There appears to be a syntax error (the `const {data}` is outside of the `render` method) in the first example. After moving the `const {data}` line into `render`, the transformed code is similar to your example, except the hoisted function definition has a `data` parameter, whereas in yours the first parameter is not `data`).

```
import { babelBind as _testBind } from \\"../../src\\";

function _testBBHoisted(data, {
  propA,
  propB
}) {
  return this.setState({
    data: {
      propA,
      propB
    }
  });
}

import * as React from \\"react\\";

function View(props) {
  return null;
}

class MyComponent extends React.Component {
  render() {
    const {
      data
    } = this.state;
    return <View onClick={_testBind(_testBBHoisted, this, data)} />;
  }
}
```

In your example, the bug is much more serious, because the transformed code will incorrectly call `setState` with the old `data` instead of the data in the click handler's first parameter.

In my test, the bug is minor, and is simply causing an unnecessary identifier to be extracted into the hoisted function parameter which might cause some excess re-rendering, but should not cause any actual behavioral bugs.

I'll put out a fix for the bug where the `data` identifier should not be pulled into the hoisted function parameters, but I'm worried that I'm still missing something because I can't reproduce the bug that you are describing.

**Comment by carbonjesse** [comment]
I'm using 0.2.2

In my error case, the data variable is actually just passed through to the hoisted function, but not used. Only the key "data" is used. So i think I have the same bug as you.

In your example, data is the first parameter. In mine it's second. Otherwise seems the same.

Yes, looking forward to a fix, thanks!

**Comment by carbonjesse** [comment]
hi @dounan have you had a chance to work on this yet? If not I might be able to give it a try.

**Comment by dounan** [comment]
I have a pr open for it. https://github.com/flexport/reflective-bind/pull/34 Let me bug the reviewer 😝

**Comment by dounan** [comment]
The PR has been merged and a new version 0.2.3 has been released!

**Comment by carbonjesse** [comment]
Awesome, thanks Dounan for sticking with it!
